infer/default_provider.go

```diff
@@ -0,0 +1,251 @@
+		dp.metadata.LanguageMap = defaultMetadata.LanguageMap
+	}
+
+	return dp
+}
+
+// WithResources adds the given custom resources to the provider.
+func (dp *DefaultProvider) WithResources(resources ...InferredResource) *DefaultProvider {
+	dp.resources = append(dp.resources, resources...)
+	return dp
+}
+
+// WithComponents adds the given components to the provider.
+func (dp *DefaultProvider) WithComponents(components ...InferredComponent) *DefaultProvider {
+	dp.components = append(dp.components, components...)
+	return dp
+}
+
+// WithFunctions adds the given functions to the provider.
+func (dp *DefaultProvider) WithFunctions(functions ...InferredFunction) *DefaultProvider {
+	dp.functions = append(dp.functions, functions...)
+	return dp
+}
+
+// WithConfig adds the given config to the provider.
+func (dp *DefaultProvider) WithConfig(config InferredConfig) *DefaultProvider {
+	dp.config = config
+	return dp
+}
+
+// WithModuleMap adds the given module map to the provider.
+func (dp *DefaultProvider) WithModuleMap(moduleMap map[tokens.ModuleName]tokens.ModuleName) *DefaultProvider {
+	dp.moduleMap = moduleMap
+	return dp
+}
+
+func (dp *DefaultProvider) WithLanguageMap(languageMap map[string]any) *DefaultProvider {
+	dp.metadata.LanguageMap = languageMap
+	return dp
+}
+
+func (dp *DefaultProvider) WithDescription(description string) *DefaultProvider {
+	dp.metadata.Description = description
+	return dp
+}
+
+func (dp *DefaultProvider) WithDisplayName(displayName string) *DefaultProvider {
+	dp.metadata.DisplayName = displayName
+	return dp
+}
+
+func (dp *DefaultProvider) WithKeywords(keywords ...string) *DefaultProvider {
+	dp.metadata.Keywords = append(dp.metadata.Keywords, keywords...)
+	return dp
+}
+
+func (dp *DefaultProvider) WithHomepage(homepage string) *DefaultProvider {
+	dp.metadata.Homepage = homepage
+	return dp
+}
+
+func (dp *DefaultProvider) WithRepository(repository string) *DefaultProvider {
+	dp.metadata.Repository = repository
+	return dp
+}
+
+func (dp *DefaultProvider) WithPublisher(publisher string) *DefaultProvider {
+	dp.metadata.Publisher = publisher
+	return dp
+}
+
+func (dp *DefaultProvider) WithLogoURL(logoURL string) *DefaultProvider {
+	dp.metadata.LogoURL = logoURL
+	return dp
+}
+
+func (dp *DefaultProvider) WithLicense(license string) *DefaultProvider {
+	dp.metadata.License = license
+	return dp
+}
+
+func (dp *DefaultProvider) WithPluginDownloadURL(pluginDownloadURL string) *DefaultProvider {
+	dp.metadata.PluginDownloadURL = pluginDownloadURL
+	return dp
+}
+
+// WithName sets the provider name.
+func (dp *DefaultProvider) WithName(name string) *DefaultProvider {
+	dp.name = name
+	return dp
+}
+
+// WithVersion sets the provider version.
+func (dp *DefaultProvider) WithVersion(version string) *DefaultProvider {
+	dp.version = version
+	return dp
+}
+
+// Build constructs the provider options based on the current state of the default provider.
+func (dp *DefaultProvider) Build() *Options {
```

Comment:
Why does this return a `*Options`? `infer.Provider` takes a `Options`.
```suggestion
func (dp *DefaultProvider) Build() Options {
```